docs/guides/optimizers.rst

```diff
@@ -4,7 +4,398 @@
+
+Using recorders with optimizers
+-------------------------------
+
+The previous section focused on using recorders as a standalone feature. We will now explore
+how recorders can be used with optimizers.
+
+If you take a closer look at the initializer of base :class:`Optimizer` class, you'll see
+that it accepts a `recorder` argument. This argument is a factory, which means it should be a
+callable that, when provided a cost function, returns a recorder. It defaults to
+the :func:`recorder <orquestra.opt.history.recorder>` that we just discussed. We can replace it with
+our own factory. The example below shows how to minimize the Rosenbrock function,
+and record its evaluation using both the default :func:`recorder` and the custom one provided by us.
+
+.. literalinclude:: ../examples/guides/optimizers.py
+  :language: python
+  :start-after: # custom recorder
+  :end-before: # --- End
+
+.. code:: text
+
+   [0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76]
+
+As can be seen, our optimizer saved only every second evaluation of the cost function when
+we passed :code:`keep_history=True`. Under the hood, the optimizer constructed an instance
+of the recorder according to the recipe that we specified.
+
+The information presented so far should be enough for optimizing arbitrary cost function
+and control how its evaluation history is stored. Let us now move to more advanced topics.
+
+Creating your own optimizer
+===========================
+
+So far we only discussed using already existing optimizers. However, you might want to implement
+an optimizer using some fancy optimization method that is not yet available in Orquestra.
+
+The way it is done is by subclassing the :class:`Optimizer` class. At bare minimum,
+your implementation has to include the :meth:`_minimize` method that implements the actual
+minimization algorithm.
+
+Optionally, you may also override the :meth:`_preprocess_cost_function` method. This method
+is responsible for two things:
+
+- validating the passed cost function, and raising an error if some problems are detected.
+- wrapping/modifying the cost function to adapt it to use with the given optimizer.
+
+For instance, the :meth:`_preprocess_cost_function` can make sure that the function has
+a gradient if the optimizer requires it. Or it can add an approximate gradient, if the
+cost function doesn't have one.
+
+The following snippet may serve as a template for implementing your own optimizer.
+
+.. literalinclude:: ../examples/guides/optimizers.py
+  :language: python
+  :start-after: # custom optimizer
+  :end-before: # --- End
+
+
+List of currently available optimizers
+======================================
+
+We conclude this guide by listing all optimizers currently available in orquestra-opt.
+
+Optimizers available with the default installation:
```

Comment:
Two thing:
1. To improve searchability, I'd also add what optimizers are available through Scipy, Qiskit, etc., at least the most popular ones.
2. Vlad added PSO recently, might be worth adding it to the list?